sensor/common/detector/baseline/baseline.go

```diff
@@ -75,16 +89,199 @@ func (w *baselineEvaluator) AddBaseline(baseline *storage.ProcessBaseline) {
 	}
 	containerNameMap[baseline.GetKey().GetContainerName()] = baselineSet
 
-	log.Infof("Successfully added process baseline %s", baseline.GetId())
+	log.Debugf("Successfully added process baseline %s", baseline.GetId())
 }
 
 // IsInBaseline checks if the process indicator is within a locked baseline
 // If the baseline does not exist, then we return true
 func (w *baselineEvaluator) IsOutsideLockedBaseline(pi *storage.ProcessIndicator) bool {
+	if pi == nil {
+		return false // Treat nil process as within baseline
+	}
+
 	w.baselineLock.RLock()
 	defer w.baselineLock.RUnlock()
 
 	baseline := w.baselines[pi.GetDeploymentId()][pi.GetContainerName()]
 	// If there is no baseline, then we are counting it as if it's within the baseline
 	return baseline != nil && !baseline.Contains(processbaseline.BaselineItemFromProcess(pi))
 }
+
+// optimizedBaselineEvaluator implements memory-optimized baseline evaluation using process set deduplication
+type optimizedBaselineEvaluator struct {
+	// deployment -> container name -> content hash (direct access)
+	deploymentBaselines map[string]map[string]string
+	// content hash -> reference count and StringSet for deduplication
+	processSets map[string]*processSetEntry
+	// lock for thread safety
+	lock sync.RWMutex
+}
+
+type processSetEntry struct {
+	refCount  int
+	processes set.StringSet
+}
+
+// newOptimizedBaselineEvaluator creates the optimized baseline evaluator implementation
+func newOptimizedBaselineEvaluator() Evaluator {
+	return &optimizedBaselineEvaluator{
+		deploymentBaselines: make(map[string]map[string]string),
+		processSets:         make(map[string]*processSetEntry),
+	}
+}
+
+// removeReference decrements reference count and cleans up if necessary
+func (oe *optimizedBaselineEvaluator) removeReference(contentHash string) {
```

Comment:
I think @vikin91 suggested splitting this off from `removeBaseline`, but I am concerned about calling one without the other.